WL-4017 Added new method in Forums to get recent messages for a site.

ouit0408 opened the pull request:
New method 'getRecentDiscussionForumMessagesByContext' added in
MessageForumsMessageManager to return given number of recent messages
for a given siteId.
Query(findRecentDiscussionForumMessagesInSite) to get recent forum messages
is added in MessageImpl.hbm.xml
New separate method 'displayMessages' added in ForumsEntityProviderImpl
to handle requests '/direct/forums/messages/SITEID.json?n=10'.

buckett commented on messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/MessageImpl.hbm.xml:
```diff
@@ -808,6 +808,17 @@
+             join topic.openForum as forum
+             join forum.area as area
+             where area.contextId = :contextId and
+             message.inReplyTo is null and
```
Won't this only get new threads in the topic, rather than any new message?

ouit0408 replied:
that is how CPD wanted it in the widget.

buckett replied:
Ok, but then the JavaDoc on the method is wrong as it doesn't get new messages it gets new threads.

buckett commented on messageforums-app/src/java/org/sakaiproject/tool/messageforums/entityproviders/ForumsEntityProviderImpl.java:
```diff
@@ -443,4 +439,58 @@ private void checkSiteAndToolAccess(String siteId) throws EntityException {
+		List<SparseMessage> messages = new ArrayList<SparseMessage>();
+
+		for(Message fm : (List<Message>) forumManager.getRecentDiscussionForumMessagesByContext(siteId, numberOfMessages)) {
+			//message has user_Id set in the 'modifiedBy' field, setting it to displayId for display purpose
```
Don't you need a permission check here as just because a user can access a site and the tool, doesn't mean they can access the message.

ouit0408 commented:
Have updated this PR and pushed the changes.

buckett commented on messageforums-app/src/java/org/sakaiproject/tool/messageforums/entityproviders/ForumsEntityProviderImpl.java:
```diff
@@ -443,4 +439,64 @@ private void checkSiteAndToolAccess(String siteId) throws EntityException {
+		for(Message fm : (List<Message>) forumManager.getRecentDiscussionForumMessagesByContext(siteId, numberOfMessages)) {
+			Topic topic = fm.getTopic();
+			//check read permission for the current user
+			if(!uiPermissionsManager.isRead(topic.getId(),((DiscussionTopic)topic).getDraft(),false, userId, siteId)) {
```
If the recent discussion has been in a folder that the user doesn't have permission to view then the recent messages will be empty.

buckett replied:
It might be worth looking at how the synoptic tool builds it's looks of new messages.

ouit0408 replied:
are you talking about the synoptic tool on home page ? it just shows the message that there are some new messages in the forum , when clicked,  goes to the forum tool.

buckett replied:
But doesn't it have a count of how many messages are new? Or does it not actually check if the user can see those messages?

buckett commented on messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/MessageImpl.hbm.xml:
```diff
@@ -808,6 +808,17 @@
+
+    <!-- returns recent forum  threads for a given siteId -->
+    <query name="findRecentDiscussionForumThreadsInSite">
+        <![CDATA[select message, topic, forum from org.sakaiproject.component.app.messageforums.dao.hibernate.MessageImpl as message
```
Performance on the query looks good. Joins have indexes, as does the inReplyTo column. The deleted column doesn't but an index wouldn't help much as most messages won't be deleted and it's a boolean.

ouit0408 commented:
Changes made and PR updated.

buckett commented on messageforums-api/src/java/org/sakaiproject/api/app/messageforums/MessageForumsMessageManager.java:
```diff
@@ -248,4 +248,12 @@ public List findPvtMsgsBySearchText(final String typeUuid, final String searchTe
\ No newline at end of file
+
+	/**
+	 * Returns a given number of recent threads for a given list of topicIds
```
Could you include the fact that it will exclude draft and deleted messages.

ouit0408 replied:
Done.

ouit0408 committed:
WL-4017 Added new method in Forums to get recent threads for a site.

New method 'getRecentDiscussionForumThreadsByTopicIds' added in
MessageForumsMessageManager to return given number of recent messages
for a given list of 'topicIds'.
Query(getRecentDiscussionForumThreadsByTopicIds) to get recent forum threads
is added in MessageImpl.hbm.xml
New separate method 'displayMessages' added in ForumsEntityProviderImpl
to handle requests '/direct/forums/messages/SITEID.json?n=10'.
